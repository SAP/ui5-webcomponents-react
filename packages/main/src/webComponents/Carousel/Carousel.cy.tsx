import { cypressPassThroughTestsFactory } from '@/cypress/support/utils';
import { Carousel } from './index';

describe('Carousel', () => {
  it('Basic Test (generated)', () => {
    cy.mount(<Carousel />);
    cy.get('[ui5-carousel]').should('exist');
  });

  cypressPassThroughTestsFactory(Carousel);
});
