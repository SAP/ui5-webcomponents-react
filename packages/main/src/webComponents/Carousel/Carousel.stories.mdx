import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { Carousel } from './index';
import { CarouselArrowsPlacement } from '../../enums/CarouselArrowsPlacement';
import { DocsHeader } from '@docs/DocsHeader';
import { Footer } from '@docs/Footer';

<Meta
  title="Layouts & Floorplans / Carousel"
  component={Carousel}
  argTypes={{
    children: { control: { disable: true } }
  }}
  args={{
    arrowsPlacement: CarouselArrowsPlacement.Content,
    style: {},
    className: '',
    tooltip: '',
    slot: '',
    ref: null
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      return (
        <Carousel {...args}>
          <img src="https://sap.github.io/ui5-webcomponents/assets/images/sample1.jpg" alt="img-sample 1" />
          <img src="https://sap.github.io/ui5-webcomponents/assets/images/sample2.jpg" alt="img-sample 2" />
          <img src="https://sap.github.io/ui5-webcomponents/assets/images/sample3.jpg" alt="img-sample 3" />
        </Carousel>
      );
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />

## Usage

When to use:

- The items you want to display are very different from each other.
- You want to display the items one after the other.
  When not to use
- The items you want to display need to be visible at the same time.
- The items you want to display are uniform and very similar For the `Carousel`.

<Footer />
