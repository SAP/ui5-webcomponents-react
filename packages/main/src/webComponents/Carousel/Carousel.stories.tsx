import type { Meta, StoryObj } from '@storybook/react';
import CarouselArrowsPlacement from '@ui5/webcomponents/dist/types/CarouselArrowsPlacement.js';
import { Carousel } from './index';

const meta = {
  title: 'Layouts & Floorplans / Carousel',
  component: Carousel,
  argTypes: {
    children: { control: { disable: true } }
  },
  args: {
    arrowsPlacement: CarouselArrowsPlacement.Content
  },
  tags: ['package:@ui5/webcomponents']
} satisfies Meta<typeof Carousel>;
export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render(args) {
    return (
      <Carousel {...args}>
        <img src="https://sap.github.io/ui5-webcomponents/images/sample1.jpg" alt="img-sample 1" />
        <img src="https://sap.github.io/ui5-webcomponents/images/sample2.jpg" alt="img-sample 2" />
        <img src="https://sap.github.io/ui5-webcomponents/images/sample3.jpg" alt="img-sample 3" />
      </Carousel>
    );
  }
};
