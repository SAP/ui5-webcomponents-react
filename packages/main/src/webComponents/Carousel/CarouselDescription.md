## Usage

### When to use:

*   The items you want to display are very different from each other.
*   You want to display the items one after the other.

### When not to use:

*   The items you want to display need to be visible at the same time.
*   The items you want to display are uniform and very similar.

## Keyboard Handling

### Basic Navigation

When the `Carousel` is focused the user can navigate between the items with the following keyboard shortcuts:  
\*

### Fast Navigation

This component provides a build in fast navigation group which can be used via `F6 / Shift + F6` or `Ctrl + Alt(Option) + Down / Ctrl + Alt(Option) + Up`. In order to use this functionality, you need to import the following module: `import "@ui5/webcomponents-base/dist/features/F6Navigation.js"`  
  

*   \[UP/DOWN\] - Navigates to previous and next item
*   \[LEFT/RIGHT\] - Navigates to previous and next item