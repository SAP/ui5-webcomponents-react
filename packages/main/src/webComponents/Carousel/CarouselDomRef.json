[
  {
    "name": "navigateTo",
    "visibility": "public",
    "since": "1.0.0-rc.15",
    "parameters": [
      { "name": "itemIndex", "type": "Integer", "optional": false, "description": "The index of the target page" }
    ],
    "description": "Changes the currently displayed page."
  }
]
