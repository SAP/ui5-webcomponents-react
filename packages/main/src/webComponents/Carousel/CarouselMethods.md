## Methods

This component exposes public methods. You can invoke them directly on the instance of the component, e.g. by using React Refs.

| Name           | Parameters                                                                                                                                      | Description                           |
| -------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------- |
| **navigateTo** | <dl><dt className="methodText">**itemIndex**</dt><dd className="methodText">The index of the target page</dd><dd><code>Integer</code></dd></dl> | Changes the currently displayed page. |
