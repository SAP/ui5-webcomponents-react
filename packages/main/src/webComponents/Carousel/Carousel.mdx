import { ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import { Canvas, Meta } from '@storybook/blocks';
import * as ComponentStories from './Carousel.stories';

<Meta of={ComponentStories} />

<DocsHeader since="0.9.0" />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<Footer />
