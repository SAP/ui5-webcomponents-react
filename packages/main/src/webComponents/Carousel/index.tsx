import { ReactNode } from 'react';
import { CarouselArrowsPlacement } from '../../enums';
import { CommonProps } from '../../interfaces/CommonProps';
import { Ui5CustomEvent } from '../../interfaces/Ui5CustomEvent';
import { Ui5DomRef } from '../../interfaces/Ui5DomRef';
import { withWebComponent } from '../../internal/withWebComponent';

import '@ui5/webcomponents/dist/Carousel.js';

interface CarouselAttributes {
  /**
   * Defines the position of arrows.
   *
   * Available options are:
   *
   * *   `Content`
   * *   `Navigation`
   *
   *
   * When set to "Content", the arrows are placed on the sides of the current page.
   * When set to "Navigation", the arrows are placed on the sides of the page indicator.
   */
  arrowsPlacement?: CarouselArrowsPlacement | keyof typeof CarouselArrowsPlacement;
  /**
   * Defines whether the carousel should loop, i.e show the first page after the last page is reached and vice versa.
   */
  cyclic?: boolean;
  /**
   * Defines the visibility of the navigation arrows. If set to true the navigation arrows will be hidden.
   *
   * **Note:** The navigation arrows are never displayed on touch devices. In this case, the user can swipe to navigate through the items.
   */
  hideNavigationArrows?: boolean;
  /**
   * Defines the visibility of the paging indicator. If set to true the page indicator will be hidden.
   */
  hidePageIndicator?: boolean;
  /**
   * Defines the number of items per page on large size (more than 1024px). One item per page shown by default.
   */
  itemsPerPageL?: number;
  /**
   * Defines the number of items per page on medium size (from 640px to 1024px). One item per page shown by default.
   */
  itemsPerPageM?: number;
  /**
   * Defines the number of items per page on small size (up to 640px). One item per page shown by default.
   */
  itemsPerPageS?: number;
}

export interface CarouselDomRef extends CarouselAttributes, Ui5DomRef {
  /**
   * Changes the currently displayed page.
   * @param {number} itemIndex - The index of the target page
   */
  navigateTo: (itemIndex: number) => void;
}

export interface CarouselPropTypes extends CarouselAttributes, CommonProps {
  /**
   * Defines the content of the component.
   */
  children?: ReactNode | ReactNode[];
  /**
   * Fired whenever the page changes due to user interaction, when the user clicks on the navigation arrows or while resizing, based on the `items-per-page-l`, `items-per-page-m` and `items-per-page-s` properties.
   */
  onNavigate?: (event: Ui5CustomEvent<HTMLElement, { selectedIndex: number }>) => void;
}

/**
 * The Carousel allows the user to browse through a set of items. The component is mostly used for showing a gallery of images, but can hold any other HTML element.
 * There are several ways to perform navigation:
 *
 * *   on desktop - the user can navigate using the navigation arrows or with keyboard shorcuts.
 * *   on mobile - the user can use swipe gestures.
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/Carousel" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const Carousel = withWebComponent<CarouselPropTypes, CarouselDomRef>(
  'ui5-carousel',
  ['arrowsPlacement', 'itemsPerPageL', 'itemsPerPageM', 'itemsPerPageS'],
  ['cyclic', 'hideNavigationArrows', 'hidePageIndicator'],
  [],
  ['navigate']
);

Carousel.displayName = 'Carousel';

Carousel.defaultProps = {
  arrowsPlacement: CarouselArrowsPlacement.Content,
  itemsPerPageL: 1,
  itemsPerPageM: 1,
  itemsPerPageS: 1
};

export { Carousel };
