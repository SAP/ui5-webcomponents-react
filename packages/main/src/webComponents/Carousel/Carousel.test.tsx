import { render } from '@shared/tests';
import { createCustomPropsTest } from '@shared/tests/utils';
import { Carousel } from '@ui5/webcomponents-react/dist/Carousel';
import React from 'react';

describe('Carousel', () => {
  test('Basic Test (generated)', () => {
    const { asFragment } = render(<Carousel />);
    expect(asFragment()).toMatchSnapshot();
  });
  createCustomPropsTest(Carousel);
});
