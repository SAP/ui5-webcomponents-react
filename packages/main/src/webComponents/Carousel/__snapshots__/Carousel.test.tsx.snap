// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Basic Test (generated) 1`] = `
<DocumentFragment>
  <ui5-carousel
    arrows-placement="Content"
    items-per-page-l="1"
    items-per-page-m="1"
    items-per-page-s="1"
    ui5-carousel=""
  />
</DocumentFragment>
`;
